stagecoach/price-pusher/price_pusher/core/listener.py

```diff
@@ -0,0 +1,244 @@
+
+        To know if they need we check:
+            - if the latest entry found in our oracle is too old,
+            - if the price deviated too much with the prices fetched from the poller
+        """
+        if self.orchestrator_prices is None:
+            raise ValueError("Orchestrator must set the prices dictionnary.")
+
+        if len(self.orchestrator_prices.keys()) == 0:
+            return False
+
+        if len(self.oracle_prices.keys()) == 0:
+            logging.error(
+                f"LISTENER {self.id} have no oracle prices at all... Sending notification."
+            )
+            return True
+
+        for pair_id, oracle_entries in self.oracle_prices.items():
+            if pair_id not in self.orchestrator_prices:
+                continue
+            for asset_type, orchestrator_entries in self.orchestrator_prices[pair_id].items():
+                if asset_type not in oracle_entries:
+                    logging.warn(
+                        f"LISTENER {self.id} miss prices in oracle entries. Sending notification."
+                    )
+                    return True
+                oracle_entry = oracle_entries[asset_type]
+                # First check if the oracle entry is outdated
+                newest_entry = self._get_most_recent_orchestrator_entry(pair_id, asset_type)
+                if self._oracle_entry_is_outdated(pair_id, oracle_entry, newest_entry):
+                    return True
+                # If not, check its deviation
+                for entry in orchestrator_entries.values():
+                    if self._new_price_is_deviating(pair_id, entry.price, oracle_entry.price):
+                        return True
+        return False
+
+    def _new_price_is_deviating(self, pair_id: str, new_price: int, oracle_price: int) -> bool:
+        """
+        Check if a new price is in the bounds allowed by the configuration.
+        """
+        max_deviation = self.price_config.price_deviation * oracle_price
+
+        lower_bound = oracle_price - max_deviation
+        upper_bound = oracle_price + max_deviation
+
+        is_deviating = not (lower_bound < new_price < upper_bound)
+        if is_deviating:
+            # TODO: show current deviation
+            logger.info(
+                f"🔔 Newest price for {pair_id} is deviating from the "
+                "config bounds. Triggering an update!"
+            )
+        return is_deviating
+
+    def _oracle_entry_is_outdated(
+        self, pair_id: str, oracle_entry: Entry, newest_entry: Entry
+    ) -> bool:
+        """
+        Check if the newest entry is recent enough to trigger an update.
+        We do that by checking the difference between the most recent entry from
+        the orchestrator and the most recent entry for the oracle.
+        """
+        max_time_elapsed = self.price_config.time_difference
+
+        delta_t = newest_entry.get_timestamp() - oracle_entry.get_timestamp()
```

Comment:
it's probably the opposite ? and names are quite confusing here, `newest_entry` corresponds to the latest pushed entry where as `oracle_entry` is the actual new entry that could be pushed

Reply:
this is the opposite:
- `oracle_entry` is the last entry we found in the oracle
- `newest_entry` is the newest entry fetched from the APIs